### PR TITLE
std.crypto.kem.kyber: mitigate KyberSlash

### PR DESCRIPTION
**Pull request by jedisct1**
On some architectures, including AMD Zen CPUs, dividing a secret by a constant denominator may not be a constant-time operation.

And most Kyber implementations, including ours, could leak the hamming weight of the shared secret because of this. See:

https://kyberslash.cr.yp.to

Multiplications aren't guaranteed to be constant-time either, but at least on the CPUs we support, they are.

### CODE REVIEW

**Comment by andrewrk** [comment]
Note that, while unlikely, this still permits the compiler to detect that the code is equivalent to a division, and lower to a division.

Proper solution will be #1776. Although we don't have a fully specified proposal, I have marked that issue as "accepted" meaning that implementing such a feature blocks the release of Zig 1.0.

**Review by andrewrk — APPROVED** [review]
Thanks!

File: lib/std/crypto/kyber_d00.zig [hunk 1]
```diff
@@ -1020,8 +1020,14 @@ const Poly = struct {
...
                 //                  = DIV((x << d) + q/2, q) & ((1<<d) - 1)
-                const t = @as(u32, @intCast(p.cs[in_off + i])) << d;
-                in[i] = @as(u16, @intCast(@divFloor(t + q_over_2, Q) & two_d_min_1));
+                const t = @as(u24, @intCast(p.cs[in_off + i])) << d;
+                // Division by invariant multiplication, equivalent to DIV(t + q/2, q).
+                // A division may not be a constant-time operation, even with a constant denominator.
+                // Here, side channels would leak information about the shared secret, see https://kyberslash.cr.yp.to
+                // Multiplication, on the other hand, is a constant-time operation on the CPUs we currently support.
+                comptime assert(((20642679 * @as(u64, Q)) >> 36) == 1);
+                const u: u32 = @intCast((@as(u64, t + q_over_2) * 20642679) >> 36);
```
Comment (bwesterb) [inline lib/std/crypto/kyber_d00.zig:1029]: Just checked this is correct for `0 ≤ t < 41520952`, viz `d ≤ 13`. Maybe we should add a static assert that `d ≤ 11` anyway.

**Comment by bwesterb** [comment]
lgtm, thanks.

File: lib/std/crypto/kyber_d00.zig (hunk 1 above)
Comment (jedisct1) [reply lib/std/crypto/kyber_d00.zig:1029]: > Just checked this is correct for `0 ≤ t < 41520952`, viz `d ≤ 13`. Maybe we should add a static assert that `d ≤ 11` anyway.

Sure, static asserts are free and never hurt!

**Pull request merged**